experimental/sql/test/ssa_to_impl/join.xdsl

```diff
@@ -0,0 +1,22 @@
+// RUN: rel_opt.py -p ssa-to-impl %s | filecheck %s
+
+module() {
+  %0 : !rel_ssa.bag<[!rel_ssa.schema_element<"a", !rel_ssa.int64>, !rel_ssa.schema_element<"b", !rel_ssa.int64>]> = rel_ssa.table() ["table_name" = "t"]
+  %1 : !rel_ssa.bag<[!rel_ssa.schema_element<"c", !rel_ssa.int64>, !rel_ssa.schema_element<"d", !rel_ssa.int64>]> = rel_ssa.table() ["table_name" = "u"]
+  %2 : !rel_ssa.bag<[!rel_ssa.schema_element<"a", !rel_ssa.int64>, !rel_ssa.schema_element<"b", !rel_ssa.int64>, !rel_ssa.schema_element<"c", !rel_ssa.int64>, !rel_ssa.schema_element<"d", !rel_ssa.int64>]> = rel_ssa.inner_join(%0 : !rel_ssa.bag<[!rel_ssa.schema_element<"a", !rel_ssa.int64>, !rel_ssa.schema_element<"b", !rel_ssa.int64>]>, %1 : !rel_ssa.bag<[!rel_ssa.schema_element<"c", !rel_ssa.int64>, !rel_ssa.schema_element<"d", !rel_ssa.int64>]>) {
+    %3 : !rel_ssa.int64 = rel_ssa.column() ["col_name" = "a"]
+    %4 : !rel_ssa.int64 = rel_ssa.column() ["col_name" = "c"]
+    %5 : !rel_ssa.bool = rel_ssa.compare(%3 : !rel_ssa.int64, %4 : !rel_ssa.int64) ["comparator" = "="]
+    rel_ssa.yield_value(%5 : !rel_ssa.bool)
+  }
+}
+
+//      CHECK:  %{{.*}} : !rel_impl.bag<[!rel_impl.schema_element<"a", !rel_impl.int64>, !rel_impl.schema_element<"b", !rel_impl.int64>]> = rel_impl.full_table_scan() ["table_name" = "t"]
+// CHECK-NEXT:  %{{.*}} : !rel_impl.bag<[!rel_impl.schema_element<"c", !rel_impl.int64>, !rel_impl.schema_element<"d", !rel_impl.int64>]> = rel_impl.full_table_scan() ["table_name" = "u"]
+// CHECK-NEXT:  %{{.*}} : !rel_impl.bag<[!rel_impl.schema_element<"a", !rel_impl.int64>, !rel_impl.schema_element<"b", !rel_impl.int64>, !rel_impl.schema_element<"c", !rel_impl.int64>, !rel_impl.schema_element<"d", !rel_impl.int64>]> = rel_impl.nested_loop_join(%{{.*}} : !rel_impl.bag<[!rel_impl.schema_element<"a", !rel_impl.int64>, !rel_impl.schema_element<"b", !rel_impl.int64>]>, %{{.*}} : !rel_impl.bag<[!rel_impl.schema_element<"c", !rel_impl.int64>, !rel_impl.schema_element<"d", !rel_impl.int64>]>) {
+// CHECK-NEXT:  ^{{.*}}(%{{.*}} : !rel_impl.tuple<[!rel_impl.schema_element<"a", !rel_impl.int64>, !rel_impl.schema_element<"b", !rel_impl.int64>]>, %{{.*}} : !rel_impl.tuple<[!rel_impl.schema_element<"c", !rel_impl.int64>, !rel_impl.schema_element<"d", !rel_impl.int64>]>):
```

Comment:
At some point, we probably want to detect equi-joins and represent them with a dedicated op instead of a generic "theta" join. I guess an equi-join would have a list of column names for the left and right side, which need to be equal pairwise. Not that only equi-joins have O(n + m) or O(log n + log m) running time; theta joins need to build the Cartesian product and test each combination of records in O(n*m) time. Detecting them is thus required before we can decide to use a hash or sort join.

Have you thought about this? Where do you think is the best place to do that? One could carry the theta joins all to way to impl, detect them in that dialect, and thus only require non-theta joins from that point on. In relational algebra, one often has a natural join, which is an equi-join with assumptions on the names, plus a theta join, but strictly speaking, only the latter is required and the former is syntactic sugar.

Reply:
On a related note, it would also be interesting to detect joins from Cartesian products + filters. That, however, I think is best done in the alg dialect.

Reply:
And the opposite: an inner join without predicates (or even with non-equi predicates) could be canonicalized to a Cartesian product (plus filters).

Reply:
At this point, I feel that it might even make sense to not have theta joins at all, but only equi-joins and Cartesian products. Maybe also natural joins, but AFAIU, these are basically inferred by their schemas and just a flavor of equi-joins (so maybe natural joins on rel_alg level lowered to equi-joins on this level?).

Reply:
So, I guess there are two ways here: a) only model Cartesian products or b) model Cartesian products and equi-joins. As both require Cartesian products, I will implement that as a first version. After that, we can still decide whether we want equi-joins for this prototype. I must admit that I don't quite understand how naming is/should be handled in Cartesian products yet. I guess I can implement another version that assumes  distinct names, but this is clearly not an end solution...

One thing that I feel a bit uneasy about is that I feel like we are losing some semantics by deciding to model joins as Cartesian product + filter. In particular, semantics that were there before and might have to be rediscovered later on. That's clearly not what out multi-level IR approach wants to do, I think it is quite literally the opposite of what we actually want to do (preserve just enough semantics through the layers to apply optimizations).

Notice that I assume that left/right outer joins are basically just other "types" of Cartesian products, so in a finalized pipeline, these would have to be represented as well.

Reply:
Yes, I agree, only having Cartesian product is quite limited for what we want to do. I didn't want to argue in favor of doing that. I do think that we should have equi-joins, i.e., go with option b).

For the name, the text book way is to say that Cartesian products of inputs with overlapping attribute names is undefined (or, as syntactic sugar, the attribute names are prefixed with the relation name on both sides, assume those are different). So I guess requiring distinctness is a sensible choice.

Outer joins are an interesting point! Note that "pure" relational algebra does not define NULL values, so it also does not have outer joins. Many extensions define them, though, but get into all problems you get into if you have NULLs. One possibility to model that are two Boolean attributes `outer_left` and `outer_right` (or similar), which is maybe more concise than having four different ops.